les/balance.go

```diff
@@ -101,6 +104,31 @@ func (bt *balanceTracker) balanceToPriority(b balance) int64 {
 	return int64(b.neg)
 }
 
+func (bt *balanceTracker) posBalanceMissing(targetPriority int64, targetCapacity uint64, after time.Duration) uint64 {
```

Comment:
Please add the function description and relative unit tests.

Reply:
```go
func (bt *balanceTracker) budget(targetPriority int64, targetCapacity uint64, after time.Duration)  uint64 {
	current := bt.balanceToPriority(bt.balance)
	if targetPriority >= current {
		if bt.negTimeFactor == 0 {
			return 0
		}
		diff :=  time.Duration(float64(targetPriority - current)/bt.negTimeFactor)
		if diff >= after {
			return 0
		}
		after -= diff
	}
	// We have to cover the connection expense by positive balance.
	if targetPriority >= 0 {
		targetPriority = -1
	}
	budget := uint64(float64(^targetPriority)*float64(targetCapacity) + float64(after)*bt.timeFactor)
	if budget >= maxBalance {
		return math.MaxUint64 // target not reachable
	}
	if budget > bt.balance.pos {
		return budget - bt.balance.pos
	}
	return 0
}
```

Reply:
Just open a question here. So we can only guarantee the idle connection by the budget. However request can also cost a lot. TBH I think request cost should be much higher than connection cost.

So should we consider the estimated request cost when we calculate the budget? Although we can never have a correct number.